// timber/dist/index.js
'use strict';

var _attach = require('./utils/attach');

var _attach2 = _interopRequireDefault(_attach);

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

var _install = require('./install');

var _install2 = _interopRequireDefault(_install);

var _middlewares = require('./middlewares');

var _middlewares2 = _interopRequireDefault(_middlewares);

var _transports = require('./transports');

var _transports2 = _interopRequireDefault(_transports);

var _formatters = require('./formatters');

var _formatters2 = _interopRequireDefault(_formatters);

var _events = require('./events');

var _events2 = _interopRequireDefault(_events);

var _contexts = require('./contexts');

var _contexts2 = _interopRequireDefault(_contexts);

var _log_entry = require('./log_entry');

var _log_entry2 = _interopRequireDefault(_log_entry);

require('./console');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// This is the main file that gets referenced by node
module.exports = {
  attach: _attach2.default,
  config: _config2.default,
  contexts: _contexts2.default,
  events: _events2.default,
  formatters: _formatters2.default,
  install: _install2.default,
  LogEntry: _log_entry2.default,
  middlewares: _middlewares2.default,
  transports: _transports2.default
};
